File: gateway/src/pricing.ts
/**
 * Per-model price table — the ONE place model costs live.
 *
 * Prices are USD per 1,000,000 tokens. Edit this constant (or override the whole
 * table via the MODEL_PRICES env var as JSON) when Anthropic pricing changes.
 * `estimateCostUsd` turns token usage into the dollar amount the budget engine
 * tracks — but the engine itself knows nothing about tokens or models, so other
 * cost sources can be priced however they like.
 */
export interface ModelPrice {
  inputPerMTok: number;
  outputPerMTok: number;
}

// Anthropic — unchanged from earlier phases.
const ANTHROPIC_PRICES: Record<string, ModelPrice> = {
  "claude-opus-4-8": { inputPerMTok: 15, outputPerMTok: 75 },
  "claude-sonnet-4-6": { inputPerMTok: 3, outputPerMTok: 15 },
  "claude-haiku-4-5-20251001": { inputPerMTok: 1, outputPerMTok: 5 },
};

// Phase 8 — OpenAI. Safe defaults (USD per 1M tokens); override the whole table
// via the MODEL_PRICES env var. Gemini/Grok tables drop in below the same way.
const OPENAI_PRICES: Record<string, ModelPrice> = {
  "gpt-4o": { inputPerMTok: 2.5, outputPerMTok: 10 },
  "gpt-4o-mini": { inputPerMTok: 0.15, outputPerMTok: 0.6 },
  "gpt-4.1": { inputPerMTok: 2, outputPerMTok: 8 },
  "gpt-4.1-mini": { inputPerMTok: 0.4, outputPerMTok: 1.6 },
  "gpt-4.1-nano": { inputPerMTok: 0.1, outputPerMTok: 0.4 },
  o1: { inputPerMTok: 15, outputPerMTok: 60 },
  "o3-mini": { inputPerMTok: 1.1, outputPerMTok: 4.4 },
  "o4-mini": { inputPerMTok: 1.1, outputPerMTok: 4.4 },
};

// Phase 8 — Google Gemini (developer API). Safe defaults (USD per 1M tokens);
// override the whole table via MODEL_PRICES. Same per-provider spread pattern.
const GOOGLE_PRICES: Record<string, ModelPrice> = {
  "gemini-2.5-pro": { inputPerMTok: 1.25, outputPerMTok: 10 },
  "gemini-2.5-flash": { inputPerMTok: 0.3, outputPerMTok: 2.5 },
  "gemini-2.5-flash-lite": { inputPerMTok: 0.1, outputPerMTok: 0.4 },
  "gemini-2.0-flash": { inputPerMTok: 0.1, outputPerMTok: 0.4 },
  "gemini-2.0-flash-lite": { inputPerMTok: 0.075, outputPerMTok: 0.3 },
  "gemini-1.5-flash": { inputPerMTok: 0.075, outputPerMTok: 0.3 },
  "gemini-1.5-pro": { inputPerMTok: 1.25, outputPerMTok: 5 },
};

// One merged table — the single lookup the budget + ranking already read. Adding
// a provider is just another spread here; no caller changes.
const DEFAULT_PRICES: Record<string, ModelPrice> = {
  ...ANTHROPIC_PRICES,
  ...OPENAI_PRICES,
  ...GOOGLE_PRICES,
};

// Fallback for any model not in the table — priced as a mid-tier model.
const FALLBACK_PRICE: ModelPrice = { inputPerMTok: 3, outputPerMTok: 15 };

/** Optional full-table override from env (JSON). Falls back to the constants. */
function loadPrices(): Record<string, ModelPrice> {
  const raw = process.env.MODEL_PRICES;
  if (!raw) return DEFAULT_PRICES;
  try {
    return { ...DEFAULT_PRICES, ...(JSON.parse(raw) as Record<string, ModelPrice>) };
  } catch {
    return DEFAULT_PRICES;
  }
}

const PRICES = loadPrices();

export function priceFor(model: string | null): ModelPrice {
  return (model && PRICES[model]) || FALLBACK_PRICE;
}

/** Dollar cost of one call given its token usage (nulls treated as 0). */
export function estimateCostUsd(
  model: string | null,
  inputTokens: number | null,
  outputTokens: number | null,
): number {
  const price = priceFor(model);
  const input = ((inputTokens ?? 0) / 1_000_000) * price.inputPerMTok;
  const output = ((outputTokens ?? 0) / 1_000_000) * price.outputPerMTok;
  return input + output;
}

File: app/pricing.py
from decimal import ROUND_HALF_UP, Decimal
from typing import NamedTuple


class Price(NamedTuple):
    input: Decimal  # dollars per million input tokens
    output: Decimal  # dollars per million output tokens


def _price(input_: str, output: str) -> Price:
    return Price(Decimal(input_), Decimal(output))


# Dollars per million tokens at first-party Anthropic API list rates.
# A model missing from this table is priced as unknown: its tokens are still
# saved, cost_usd stays null.
PRICES: dict[str, Price] = {
    "claude-fable-5": _price("10.00", "50.00"),
    "claude-mythos-5": _price("10.00", "50.00"),
    "claude-opus-5": _price("5.00", "25.00"),
    "claude-opus-4-8": _price("5.00", "25.00"),
    "claude-opus-4-7": _price("5.00", "25.00"),
    "claude-opus-4-6": _price("5.00", "25.00"),
    # Sonnet 5 carries a $2.00/$10.00 introductory rate through 2026-08-31.
    # We bill the standard rate so the table stays date-independent.
    "claude-sonnet-5": _price("3.00", "15.00"),
    "claude-sonnet-4-6": _price("3.00", "15.00"),
    "claude-haiku-4-5": _price("1.00", "5.00"),
    "claude-haiku-4-5-20251001": _price("1.00", "5.00"),
}

PER_MILLION = Decimal(1_000_000)
# Six decimals: a single cheap request can cost well under a cent.
RESOLUTION = Decimal("0.000001")


def cost_usd(
    model: str | None, input_tokens: int | None, output_tokens: int | None
) -> Decimal | None:
    """Dollar cost of one request, or None when the model or the usage is unknown."""
    price = PRICES.get(model) if model else None
    if price is None:
        return None
    if input_tokens is None and output_tokens is None:
        return None

    total = (price.input * (input_tokens or 0) + price.output * (output_tokens or 0)) / PER_MILLION
    return total.quantize(RESOLUTION, rounding=ROUND_HALF_UP)
